<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Single-cell Expression Integration System for Mapping genetically Implicated Cell types â€¢ seismicGWAS</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Single-cell Expression Integration System for Mapping genetically Implicated Cell types">
<meta name="description" content="Discover cell type-disease associations using single-cell RNA-seq and GWAS summary statistics.">
<meta property="og:description" content="Discover cell type-disease associations using single-cell RNA-seq and GWAS summary statistics.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">seismicGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="index.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="articles/seismicGWAS.html">Introduction</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-data-processing" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Data processing</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-data-processing">
<li><a class="dropdown-item" href="articles/GWAS_processing.html">GWAS preprocessing</a></li>
    <li><a class="dropdown-item" href="articles/scRNA-seq_processing.html">scRNA-seq preprocessing</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ylaboratory/seismic/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="seismic-single-cell-expression-integration-system-for-mapping-genetically-implicated-cell-types">
<em>seismic</em>: Single-cell Expression Integration System for Mapping genetically Implicated Cell types<a class="anchor" aria-label="anchor" href="#seismic-single-cell-expression-integration-system-for-mapping-genetically-implicated-cell-types"></a>
</h1></div>
<p>This repository contains code for the R package <a href="https://github.com/ylaboratory/seismic" class="external-link"><em>seismicGWAS</em></a>, an implementation of the <em>seismic</em> computational framework for mapping genetically implicated cell types and their driver genes in complex traits and diseases, by integrating GWAS summary statistics and single cell expression data.</p>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>If you use <em>seismicGWAS</em> we ask that you cite our paper:</p>
<blockquote>
<p>Disentangling associations between complex traits and cell types with <em>seismic</em>. Lai Q, Dannenfelser R, Roussarie JP, Yao V. BioRxiv. April 2024.</p>
</blockquote>
</div>
<div class="section level2">
<h2 id="about">About<a class="anchor" aria-label="anchor" href="#about"></a>
</h2>
<p>Integrating single-cell RNA sequencing (scRNA-seq) with Genome-Wide Association Studies (GWAS) can help reveal GWAS-associated cell types furthering our understanding of the cell-type-specific biological processes underlying complex traits and disease. In order to rapidly and accurately pinpoint associations, we develop a novel framework, <em>seismic</em>, which characterizes cell types using a new specificity score. As part of the <em>seismic</em> framework, the specific genes driving cell type-trait associations can easily be accessed and analyzed, enabling further biological insights. The following figure depicts a high level overview of this process.</p>
<figure><img src="reference/figures/seismic_overview.png" alt=""><figcaption><b>Overview of seismic.</b> <i>seismic</i> allows for a flexible set of cell type labels or granularities, ranging from broader cell classes to specifically defined cell types, before calculating a novel gene specificity score for each of the corresponding cell characterizations, capturing both the magnitude and consistency of gene expression. These cell-type-level gene specificity scores are then integrated with MAGMA z-scores using a regression model to assess the statistical significance of cell type-trait associations, under the assumption that disease-critical genes are more specific to the implicated cell type.
</figcaption></figure>
</div>
<div class="section level2">
<h2 id="installation-and-dependencies">Installation and dependencies<a class="anchor" aria-label="anchor" href="#installation-and-dependencies"></a>
</h2>
<p>To install the <a href="https://github.com/ylaboratory/seismic" class="external-link">seismicGWAS</a> R package first clone the <em>seismic</em> repo and then use devtools within R to point to <em>seismic</em> and install. The R environment version should be 4.0.0 or higher.</p>
<p>First download the repository to your local machine, and then directly install it in R:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="va">path_to_seismic_folder</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/ylaboratory/seismic" class="external-link">'seismicGWAS'</a></span><span class="op">)</span></span></code></pre></div>
<p>Or alternatively, you can install the package directly from GitHub using the following command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ylaboratory/seismicGWAS"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/ylaboratory/seismic" class="external-link">'seismicGWAS'</a></span><span class="op">)</span></span></code></pre></div>
<p>The package requires several dependencies which are listed in the <code>DESCRIPTION</code> file, which will be installed automatically.</p>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>Below we quickly show how to use <em>seismicGWAS</em> to calcuate cell type-trait associations for the sample data included in the package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate cell type specificity scores using included sample data</span></span>
<span><span class="co"># estimated running time: less than 1 minute</span></span>
<span><span class="va">tmfacs_sscore</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/calc_specificity.html">calc_specificity</a></span><span class="op">(</span><span class="va">tmfacs_sce_small</span>, ct_label_col<span class="op">=</span><span class="st">'cluster_name'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># convert mouse gene identifiers to human ones that match data in GWAS summary data</span></span>
<span><span class="co"># from MAGMA</span></span>
<span><span class="co"># estimated running time: 3 seconds</span></span>
<span><span class="va">tmfacs_sscore_hsa</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/translate_gene_ids.html">translate_gene_ids</a></span><span class="op">(</span><span class="va">tmfacs_sscore</span>, from<span class="op">=</span><span class="st">'mmu_symbol'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate cell type-trait associations for type 2 diabetes</span></span>
<span><span class="co"># estimated running time: less than 1 minutes</span></span>
<span><span class="fu"><a href="reference/get_ct_trait_associations.html">get_ct_trait_associations</a></span><span class="op">(</span><span class="va">tmfacs_sscore_hsa</span>, <span class="va">t2d_magma</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># find the influential genes for a significant cell type-trait association</span></span>
<span><span class="co"># in type 2 diabetes</span></span>
<span><span class="co"># estimated running time: 10 seconds</span></span>
<span><span class="fu"><a href="reference/find_inf_genes.html">find_inf_genes</a></span><span class="op">(</span><span class="st">"Pancreas.beta cell"</span>, <span class="va">tmfacs_sscore_hsa</span>, <span class="va">t2d_magma</span><span class="op">)</span></span></code></pre></div>
<p>For full usage details, including a walk through of all major functions and the extra upstream data preprocessing instructions, please check out our more extensive tutorial:</p>
<ul>
<li><a href="https://ylaboratory.github.io/seismic/articles/seismicGWAS.html">Introduction and walk through</a></li>
<li><a href="https://ylaboratory.github.io/seismic/articles/GWAS_processing.html">GWAS preprocessing</a></li>
<li><a href="https://ylaboratory.github.io/seismic/articles/scRNA-seq_processing.html">scRNA-seq preprocessing</a></li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/ylaboratory/seismic/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/ylaboratory/seismic/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/BSD-3-Clause" class="external-link">BSD_3_clause</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing seismicGWAS</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Qiliang Lai <br><small class="roles"> Author, maintainer </small>   </li>
<li>Ruth Dannenfelser <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-8766-6424" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Vicky Yao <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-3201-9983" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qiliang Lai, Ruth Dannenfelser, Vicky Yao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
