<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute seismic specificity score for each gene and cell type. — calc_specificity • seismicGWAS</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute seismic specificity score for each gene and cell type. — calc_specificity"><meta name="description" content="Compute seismic specificity score for each gene and cell type."><meta property="og:description" content="Compute seismic specificity score for each gene and cell type."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">seismicGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/seismicGWAS.html">Introduction</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-data-processing" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Data processing</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-data-processing"><li><a class="dropdown-item" href="../articles/GWAS_processing.html">GWAS preprocessing</a></li>
    <li><a class="dropdown-item" href="../articles/scRNA-seq_processing.html">scRNA-seq preprocessing</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ylaboratory/seismic/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Compute seismic specificity score for each gene and cell type.</h1>
      <small class="dont-index">Source: <a href="https://github.com/ylaboratory/seismic/blob/master/R/calc_specificity.R" class="external-link"><code>R/calc_specificity.R</code></a></small>
      <div class="d-none name"><code>calc_specificity.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Compute seismic specificity score for each gene and cell type.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calc_specificity</span><span class="op">(</span></span>
<span>  <span class="va">sce</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  ct_label_col <span class="op">=</span> <span class="st">"idents"</span>,</span>
<span>  min_uniq_ct <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  min_ct_size <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  min_cells_gene_exp <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  min_avg_exp_ct <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sce">sce<a class="anchor" aria-label="anchor" href="#arg-sce"></a></dt>
<dd><p>A SingleCellExperiment object. This object needs to include the
assay_name specified and a ct_label_col column corresponding to cell type labels
for the granularity of interest. Row names are used as the gene name identifiers.</p></dd>


<dt id="arg-assay-name">assay_name<a class="anchor" aria-label="anchor" href="#arg-assay-name"></a></dt>
<dd><p>An assay in sce, default 'logcounts' (calculated using scran).</p></dd>


<dt id="arg-ct-label-col">ct_label_col<a class="anchor" aria-label="anchor" href="#arg-ct-label-col"></a></dt>
<dd><p>A column name in colData of sce (information for the cells in sce)
that provides labels for each cell, indicating membership in each cell type
at the granularity of interest.</p></dd>


<dt id="arg-min-uniq-ct">min_uniq_ct<a class="anchor" aria-label="anchor" href="#arg-min-uniq-ct"></a></dt>
<dd><p>The minimum number of unique cell types needed in the data to
proceed with specificity score calculations. Default: 2.</p></dd>


<dt id="arg-min-ct-size">min_ct_size<a class="anchor" aria-label="anchor" href="#arg-min-ct-size"></a></dt>
<dd><p>Filter for cell types to be included in the specificity score calculation
based on the number of cells that are labeled as that cell type. Default: 20.</p></dd>


<dt id="arg-min-cells-gene-exp">min_cells_gene_exp<a class="anchor" aria-label="anchor" href="#arg-min-cells-gene-exp"></a></dt>
<dd><p>Filter for genes to be included in the specificity score calculation
based on the number of cells where the gene has non-zero expression. Default: 10.</p></dd>


<dt id="arg-min-avg-exp-ct">min_avg_exp_ct<a class="anchor" aria-label="anchor" href="#arg-min-avg-exp-ct"></a></dt>
<dd><p>Filter for genes to be included in the specificity score calculation
based on the mean of cell-type-averaged expressions (aka mean of mean cell type expression). Default: 0.1</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A dgeMatrix of specificity scores (genes as rows, cell types as columns).</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qiliang Lai, Ruth Dannenfelser, Vicky Yao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

