<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>GWAS preprocessing for _seismic_ input • seismicGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GWAS preprocessing for _seismic_ input">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">seismicGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/seismicGWAS.html">Introduction</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-data-processing" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Data processing</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-data-processing">
<li><a class="dropdown-item" href="../articles/GWAS_processing.html">GWAS preprocessing</a></li>
    <li><a class="dropdown-item" href="../articles/scRNA-seq_processing.html">scRNA-seq preprocessing</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ylaboratory/seismic/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>GWAS preprocessing for _seismic_ input</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ylaboratory/seismic/blob/master/vignettes/GWAS_processing.Rmd" class="external-link"><code>vignettes/GWAS_processing.Rmd</code></a></small>
      <div class="d-none name"><code>GWAS_processing.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <em>seismic</em> framework takes the output from <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004219" class="external-link">MAGMA</a>
as input. Basically, the MAGMA software digests SNP-level summary
statistics of a trait in a genome-wide association study into a
gene-level MAGMA z-score statistics. This tutorial will guide you
through the preprocessing steps necessary to convert raw GWAS summary
statistics into the proper format for <em>seismic</em> analysis.</p>
<p>Before beginning, ensure you have installed the MAGMA software and
obtained the required files:</p>
<ul>
<li>Gene location file</li>
<li>SNP synonyms file (dbSNP)</li>
<li>Population-specific reference data</li>
</ul>
<p>These files can be all downloaded from the <a href="https://cncr.nl/research/magma/" class="external-link">MAGMA website</a>.</p>
<p>In addition to these reference files, a GWAS summary statistics table
is also required to provide SNP-level information for a trait of
interest. We will use example files to demonstrate the preprocessing
steps. These files are available for download from our data repository
<a href="https://zenodo.org/records/15582078" class="external-link">here</a>. We will assume
that users wanting to follow this tutorial have downloaded and unzipped
the <code>all_data</code> folder in this repository before
beginning.</p>
</div>
<div class="section level2">
<h2 id="step-1-clean-up-the-gwas-summary-statistics-file">Step 1: Clean up the GWAS summary statistics file<a class="anchor" aria-label="anchor" href="#step-1-clean-up-the-gwas-summary-statistics-file"></a>
</h2>
<p>A starting GWAS summary statistics file should at minimum include
columns for: chromsome ID, base-pair position, and p-value.</p>
<p>Additionally, we strongly recommend including SNP IDs to facilitate
MAGMA analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">example_gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"all_data/example_data/sample_gwas.txt"</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">example_gwas</span><span class="op">)</span></span></code></pre></div>
<pre class="console"><code><span><span class="co">## # A tibble: 6 × 7</span></span>
<span><span class="co">##   Allele1 Allele2 Direction P.value  group Chromosome     BP</span></span>
<span><span class="co">##   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;       &lt;dbl&gt;  &lt;int&gt;      &lt;int&gt;  &lt;int&gt;</span></span>
<span><span class="co">## 1 t       c       ---        0.0541 184129          1 796338</span></span>
<span><span class="co">## 2 a       g       +++        0.0315  46089          1 817341</span></span>
<span><span class="co">## 3 t       c       +++        0.0159 189219          1 817514</span></span>
<span><span class="co">## 4 a       t       -+-        0.199   39112          1 870176</span></span>
<span><span class="co">## 5 c       g       --+        0.730   80434          1 928622</span></span>
<span><span class="co">## 6 t       c       --+        0.913  185512          1 996120</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-2-map-the-snp-location-to-grch37">Step 2: Map the SNP location to GRCh37<a class="anchor" aria-label="anchor" href="#step-2-map-the-snp-location-to-grch37"></a>
</h2>
<p>For MAGMA gene analysis, the required population-specific reference
data is based on the GRCh37 genome build, all SNP location should be
converted to GRCh37 (hg19). We illustrate the steps of this conversion
using <a href="https://genome.ucsc.edu/cgi-bin/hgLiftOver" class="external-link">LiftOver</a>
provided by UCSC. We start with hg38 data, but if your data is already
hg19 you can skip this step. Additionally, this preprocessed version is
already saved in the downloadable example files:
<code>all_data/example_data/sample_gwas_hg19.txt</code>.</p>
<p>First the chain object <code>hg38ToHg19.over.chain.gz</code> must be
downloaded from the <a href="https://hgdownload.cse.ucsc.edu/goldenpath/hg19/liftOver/" class="external-link">UCSC
website</a> and unzipped.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">'GenomicRanges'</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/IRanges" class="external-link">'IRanges'</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">'rtracklayer'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">chainObject</span> <span class="op">&lt;-</span> <span class="fu">import.chain</span><span class="op">(</span><span class="st">"hg38ToHg19.over.chain"</span><span class="op">)</span> <span class="co"># update with path to chain obj</span></span>
<span><span class="va">grObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="va">example_gwas</span><span class="op">$</span><span class="va">Chromosome</span><span class="op">)</span>,</span>
<span>                   ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">example_gwas</span><span class="op">$</span><span class="va">BP</span><span class="op">)</span>, end<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">example_gwas</span><span class="op">$</span><span class="va">BP</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hg19_res</span> <span class="op">&lt;-</span> <span class="fu">liftOver</span><span class="op">(</span><span class="va">grObject</span>, <span class="va">chainObject</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract the result data frame, select the necessary columns</span></span>
<span><span class="va">hg19_res_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">hg19_res</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="va">seqnames</span>, pattern <span class="op">=</span> <span class="st">"chr"</span>,replacement <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">group</span>, <span class="va">start</span>, <span class="va">seqnames</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rename</a></span><span class="op">(</span><span class="st">"start"</span> <span class="op">=</span> <span class="st">"BP"</span>, <span class="st">"seqnames"</span> <span class="op">=</span> <span class="st">"Chromosome"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the 1 to 1 mapping</span></span>
<span><span class="va">hg19_res_df</span> <span class="op">&lt;-</span> <span class="va">hg19_res_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># merge with the original data frame (the group column indicates the original position in the table)</span></span>
<span><span class="va">example_gwas</span> <span class="op">&lt;-</span> <span class="va">example_gwas</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>group <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Chromosome</span>, <span class="op">-</span><span class="va">BP</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">hg19_res_df</span> , by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">BP</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">example_gwas</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>BP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">BP</span><span class="op">)</span><span class="op">)</span>,  <span class="co">#avoid numerical output </span></span>
<span>            file <span class="op">=</span> <span class="st">"sample_gwas_hg19.txt"</span>, </span>
<span>            sep<span class="op">=</span><span class="st">"\t"</span>, col.names <span class="op">=</span> <span class="cn">T</span>, row.names <span class="op">=</span> <span class="cn">F</span>, quote <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">example_gwas</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-3-map-the-snp-location-to-snp-id">Step 3: Map the SNP location to SNP ID<a class="anchor" aria-label="anchor" href="#step-3-map-the-snp-location-to-snp-id"></a>
</h2>
<p>If the SNP ID is not included in the GWAS summary statistics file, we
can use the SNP synonyms file to map the SNP location to SNP ID. To
simplify the procedure, we have included a bash script in the <a href="https://github.com/ylaboratory/seismic-analysis" class="external-link">seismic-analysis
repository</a> at <code>src/tools/annotate_SNP.sh</code>. This script
will require the dbSNP file and the GWAS summary statistics file as
input, and output a new file with SNP ID included in the last column. To
view all specific parameters of the script, you can run the following
command in the terminal:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">bash</span> src/tools/annotate_SNP.sh <span class="at">-h</span></span></code></pre></div>
<pre class="console"><code><span><span class="co">## </span></span>
<span><span class="co">##     This script help you locate SNP and transform the locus (by chromosome id and location) to the corresponding variant rsid.</span></span>
<span><span class="co">##     Only intersected SNPs will be printed out.</span></span>
<span><span class="co">##     Note!!!! 1. All file path should be the same as the original files. 2. bedtools is required, so make sure it's in your environment or path</span></span>
<span><span class="co">##     Here are the required parameters:</span></span>
<span><span class="co">##     -g | GWAS summary statistics files. Or any other files that you would like to annotate SNPs to.</span></span>
<span><span class="co">##     -o | Outpur file header. If the value is given as 'dir1/dir2/output', then the final output file will be 'dir1/dir2/output.annot.table.txt'. If it is not specified, the output file will be the same as the input file with the suffix ignored.</span></span>
<span><span class="co">##     -c | Chromosome column, i.e. the column with with chromosome information be (in the table). The value should be specified or by default it will be 1.</span></span>
<span><span class="co">##     -p | Position column, i.e. The column with the the position information (in the table), The value should be specified or by default it will be 2.</span></span>
<span><span class="co">##     -b | Basis of chromosome numbering, to specify the if first position on a chromosome is labeled as '1'(1-base), or '0'(0-base). The default parameters are start with 1. #only these two are available!</span></span>
<span><span class="co">##     -k | Keep all unmapped SNPs which do not exist in the dbSNP file or not. By default it is True. (You can choose between 'True' or 'False')</span></span>
<span><span class="co">##     </span></span>
<span><span class="co">##     optional parameters:</span></span>
<span><span class="co">##     -v | VCF file for the dbsnp version you used. For default it will be located in 'data/ref/Hg19.dbsnp151.vcf'. This can be downloaded online.</span></span>
<span><span class="co">##     -n | Number of tasks to split the job. The default value is 1. If the memory is not enough, you can split the job into several parts but it will take more time.</span></span>
<span><span class="co">##     -t | Customized bedtools command. By default it is 'bedtools'. If you do not have it in your environment, you can specify the path to the bedtools command.</span></span>
<span><span class="co">##     -w | The working directory. By default it is the current directory. If you want to specify the working directory, you can use this parameter.</span></span>
<span><span class="co">##     -h | Help message.</span></span>
<span><span class="co">## </span></span></code></pre>
<p>To annotate locations, this script requires a vcf file for the dbSNP,
which is available for download from UCSC, and requires <a href="https://bedtools.readthedocs.io/en/latest/" class="external-link">bedtools</a>. Once
these prerequisties are met you can run the script as follows:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">bash</span> src/tools/annotate_SNP.sh <span class="at">-g</span> all_data/example_data/sample_gwas_hg19.txt <span class="at">-o</span> all_data/example_data/sample_gwas_hg19 <span class="dt">\</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>-c 6 <span class="at">-p</span> 7 <span class="at">-v</span> data/ref/Hg19.dbsnp151.vcf <span class="at">-t</span> bin/bedtools</span></code></pre></div>
<p>Note that the script may takes hours to run and consumes more than
100 GB of memory. To alleviate memory usage but increase the running
time, the <code>-n</code> option may be used to sequentially split the
task.</p>
<p>The resulting annotation table will be printed with the path and
header specified by the <code>-o</code> option.</p>
<pre class="console"><code>&gt; head all_data/example_data/sample_gwas_hg19.annot.table

## Allele1  Allele2 Direction   P.value group   Chromosome  BP  marker.name
## t    c   --- 0.0541  1   1   731718 rs58276399
## a    g   +++ 0.03149 2   1   752721 rs3131972
## t    c   +++ 0.01587 3   1   752894 rs3131971
## a    t   -+- 0.1987  4   1   805556 rs72631880
## c    g   --+ 0.7303  5   1   864002 rs1806501
## t    c   --+ 0.913   6   1   931500 rs74718486
## a    g   --+ 0.4902  7   1   938125 rs2710868
## a    g   --+ 0.785   8   1   941334 rs57683598
## t    c   +++ 0.1863  9   1   948921 rs15842</code></pre>
</div>
<div class="section level2">
<h2 id="step-4-generate-the-input-file-for-magma">Step 4: Generate the input file for MAGMA<a class="anchor" aria-label="anchor" href="#step-4-generate-the-input-file-for-magma"></a>
</h2>
<p>The MAGMA software takes two input files generated from the
previously processed (or unprocessed but with the required columns) GWAS
summary statistics file: a SNP location file and a p-value file. You can
simply use the <code>awk</code> command to extract the information in
the correct format in the command line:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># print rsID, chromosome, BP but not colum names:</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'{if (NR&gt;1) print $8, $6, $7}'</span> all_data/example_data/sample_gwas_hg19.annot.table <span class="op">&gt;</span> all_data/example_data/sample_gwas_hg19.snp_loc.txt</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="co"># print rsID, p-value but not colum names:</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'{if (NR&gt;1) print $8, $4}'</span> all_data/example_data/sample_gwas_hg19.annot.table <span class="op">&gt;</span> all_data/example_data/sample_gwas_hg19.p_val.txt</span></code></pre></div>
<p>or in R:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anno_tbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"all_data/example_data/sample_gwas_hg19.annot.table"</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">anno_tbl</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">marker.name</span>, <span class="va">Chromosome</span>, <span class="va">BP</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"all_data/example_data/sample_gwas_hg19.snp_loc.txt"</span>, col.names <span class="op">=</span> <span class="cn">F</span>, row.names <span class="op">=</span> <span class="cn">F</span>, quote <span class="op">=</span> <span class="cn">F</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span><span class="va">anno_tbl</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">marker.name</span>, <span class="va">P.value</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"all_data/example_data/sample_gwas_hg19.p_val.txt"</span>, col.names <span class="op">=</span> <span class="cn">F</span>, row.names <span class="op">=</span> <span class="cn">F</span>, quote <span class="op">=</span> <span class="cn">F</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span></code></pre></div>
<p>This results in a file that looks like:</p>
<pre class="console"><code>&gt; head all_data/example_data/sample_gwas_hg19.snp_loc.txt

## rs58276399   1   731718
## rs3131972    1   752721
## rs3131971    1   752894
## rs72631880   1   805556
## rs1806501    1   864002
## rs74718486   1   931500
## rs2710868    1   938125
## rs57683598   1   941334
## rs15842  1   948921
## rs9442364    1   970215</code></pre>
</div>
<div class="section level2">
<h2 id="step-5-run-magma-software">Step 5: Run MAGMA software<a class="anchor" aria-label="anchor" href="#step-5-run-magma-software"></a>
</h2>
<p>There are two steps: gene annotation and gene analysis. You can use
our implemented bash script
<code>src/tools/magma_gene_zscore_analysis.sh</code> in the <a href="https://github.com/ylaboratory/seismic-analysis" class="external-link">seismic-analysis
repository</a> to run the two steps together.</p>
<p>Another parameter to specify is the gene analysis window size, which
is set as 35 kilobase (kb) upstream to 10 kb downstream by default, as
recommend by <a href="https://www.nature.com/articles/s41588-020-0610-9" class="external-link">a previous
study</a>. Although we have shown that output files from various window
sizes make only a minor difference in the results of <em>seismic</em>
analysis, you can still change the window size by specifying the
parameter <code>--window-size</code> in the script.</p>
<p>To view all specific parameters of the script, you can run the
following command in the terminal:</p>
<pre class="console"><code>&gt; bash src/tools/magma_gene_zscore_analysis.sh -h

## Usage: magma_gene_zscore_analysis.sh -l SNP_LOC_FILE -p SNP_P_VALUE_FILE -o OUTPUT_DIR -s COHORT_SIZE -m MAGMA_PATH -g GENE_LOC_FILE -b B_FILE [-w WINDOW_SIZES] [-h]
## 
## This script performs MAGMA annotation and analysis on SNP data.
## 
## Required parameters:
##     -l | SNP location file
##     -p | SNP p-value file
##     -o | Output directory
##     -s | Cohort size (use 'cX' format to specify column X containing sample size)
##     -m | Path to MAGMA binary
##     -g | Path to gene location file
##     -b | Path to g1000_eur files (without extension)
## 
## Optional parameters:
##     -w | Window sizes (default: 35,10). Use colon-separated list for multiple sizes, e.g., '35,10:15,20:40,10'
##     -h | Display this help message</code></pre>
<p>Then finally to run MAGMA with its accompanying <a href="https://cncr.nl/research/magma/" class="external-link">reference data</a>:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">bash</span> src/tools/magma_gene_zscore_analysis.sh <span class="at">-l</span> all_data/example_data/sample_gwas_hg19.snp_loc.txt <span class="dt">\</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>-p all_data/example_data/sample_gwas_hg19.p_val.txt <span class="dt">\</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>-o all_data/example_data/ <span class="dt">\</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>-s 100000 <span class="at">-m</span> bin/magma/magma <span class="at">-g</span> ref/magma/NCBI37.3/NCBI37.3.gene.loc <span class="at">-b</span> ref/magma/g1000_eur/g1000_eur </span></code></pre></div>
<p>Besides specifying these reference file paths and the curated files
from the summary statistics, the total size of the cohort should also be
specified. Note that sometimes the cohort size for each SNP may be
different, for example, due to some quality control procedure. In that
case, a separate column in the p-value file should be added and the
input <code>-s</code> argument should tell the script the order of the
column (for example, the third column) describing the cohort size. By
specifying these arguments, the gene-level MAGMA z-score will be printed
out in the output directory with header similar to the snp location
file. In our example the output file will be:
<code>all_data/example_data/sample_gwas_hg19.35.10.genes.out</code>. In
addition, please remember to check the MAGMA running log. Sometimes when
the data are not in the correct format, there will be very few genes and
SNPs annotated (e.g., less than 10% of the SNPs).</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qiliang Lai, Ruth Dannenfelser, Vicky Yao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
