<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>scRNA-seq data preprocessing for _seismic_ input • seismicGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="scRNA-seq data preprocessing for _seismic_ input">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">seismicGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/seismicGWAS.html">Introduction</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-data-processing" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Data processing</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-data-processing">
<li><a class="dropdown-item" href="../articles/GWAS_processing.html">GWAS preprocessing</a></li>
    <li><a class="dropdown-item" href="../articles/scRNA-seq_processing.html">scRNA-seq preprocessing</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ylaboratory/seismic/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>scRNA-seq data preprocessing for _seismic_ input</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ylaboratory/seismic/blob/master/vignettes/scRNA-seq_processing.Rmd" class="external-link"><code>vignettes/scRNA-seq_processing.Rmd</code></a></small>
      <div class="d-none name"><code>scRNA-seq_processing.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <em>seismic</em> framework takes a SingleCellExperiment object as
input while requiring a column of cell metadata to specify the analysis
granularity. Before running a <em>seismic</em> analysis, the scRNAseq
data should be preprocessed for optimal usage of the <em>seismic</em>
framework. We illustrate this process with an example below with many
tips adapted from <a href="https://www.sc-best-practices.org/preprocessing_visualization/normalization.html" class="external-link">single-cell
best practices</a>.</p>
</div>
<div class="section level2">
<h2 id="step-1-load-scrna-seq-data-and-accompanying-metadata">Step 1: Load scRNA-seq data and accompanying metadata<a class="anchor" aria-label="anchor" href="#step-1-load-scrna-seq-data-and-accompanying-metadata"></a>
</h2>
<p>For simplicity, we assume that we already have the scRNA-seq data and
metadata in the form of a SingleCellExperiment object. The sample data
object used here can be downloaded from the <em>seismic</em> <a href="https://zenodo.org/records/15362163" class="external-link">zenodo data
repository</a>.</p>
<p>We will assume that users wanting to follow this tutorial have
downloaded and unzipped the <code>all_data</code> folder in before
beginning.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">example_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'all_data/example_data/sample_expr.rds'</span><span class="op">)</span> <span class="co"># can read in your own SCE file to replace</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-2-assess-data-quality">Step 2: Assess data quality<a class="anchor" aria-label="anchor" href="#step-2-assess-data-quality"></a>
</h2>
<p>We check data quality by looking at the distribution of gene
expression and the number of genes expressed in each cell. Some widely
used metric to evaluate the data quality includes: sequencing depth,
number of genes detected, mitochondrial gene percentage, etc.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># check total counts</span></span>
<span><span class="va">example_sce</span><span class="op">$</span><span class="va">tot_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">example_sce</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check total mitrochondrial counts</span></span>
<span><span class="va">example_sce</span><span class="op">$</span><span class="va">mito_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">example_sce</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^Mt-"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="va">example_sce</span><span class="op">$</span><span class="va">tot_counts</span></span>
<span></span>
<span><span class="co"># check the number of genes detected</span></span>
<span><span class="va">example_sce</span><span class="op">$</span><span class="va">detected_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">example_sce</span>, <span class="st">"counts"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Typically mitochondrial genes will be named with “Mt-” prefix in
mouse and “MT-” prefix in human data. In our data these genes are not
present. Thus, we only filter cells based on the total counts and the
number of detected genes. Additionally, in this dataset we have cell
type annotations - since these are an important part of detecting
associations we remove cells without such information.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter by total counts and detected genes</span></span>
<span><span class="va">example_sce</span> <span class="op">&lt;-</span> <span class="va">example_sce</span><span class="op">[</span>, <span class="va">example_sce</span><span class="op">$</span><span class="va">tot_counts</span> <span class="op">&gt;</span> <span class="fl">2000</span> <span class="op">&amp;</span> <span class="va">example_sce</span><span class="op">$</span><span class="va">detected_genes</span> <span class="op">&gt;</span> <span class="fl">2000</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># filter by cell ontology information</span></span>
<span><span class="va">example_sce</span> <span class="op">&lt;-</span> <span class="va">example_sce</span><span class="op">[</span>,<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">$</span><span class="va">cell_ontology_id</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">example_sce</span></span></code></pre></div>
<pre class="console"><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 23341 7346 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(23341): 0610005C13Rik 0610007C21Rik ... l7Rn6</span></span>
<span><span class="co">##   zsGreen-transgene</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(42): nReads orig.ident ... mito_ratio detected_genes</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: RNA</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<div class="section level3">
<h3 id="optional-step-remove-genes-with-low-expression-to-alleviate-memory-and-computation-burden">Optional Step: Remove genes with low expression to alleviate memory
and computation burden<a class="anchor" aria-label="anchor" href="#optional-step-remove-genes-with-low-expression-to-alleviate-memory-and-computation-burden"></a>
</h3>
<p>We can also remove genes with low expression, since these genes are
less likely to be informative for downstream analysis.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># number of cells expressing each gene</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span><span class="op">$</span><span class="va">num_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">example_sce</span> ,<span class="st">"counts"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># number of total counts deteced for each gene</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span><span class="op">$</span><span class="va">num_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">example_sce</span> ,<span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">example_sce</span> <span class="op">&lt;-</span> <span class="va">example_sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span><span class="op">$</span><span class="va">num_cells</span> <span class="op">&gt;=</span> <span class="fl">5</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span><span class="op">$</span><span class="va">num_counts</span> <span class="op">&gt;=</span> <span class="fl">10</span>,<span class="op">]</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="step-3-normalize-the-data">Step 3: Normalize the data<a class="anchor" aria-label="anchor" href="#step-3-normalize-the-data"></a>
</h2>
<p>As recommended by <a href="https://www.sc-best-practices.org/preprocessing_visualization/normalization.html" class="external-link">single-cell
best practices</a>, we then use <a href="https://bioconductor.org/packages/release/bioc/html/scran.html" class="external-link">scran</a>
to estimate the per-cell normalization factor based on cell pooling.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cell_pooling</span> <span class="op">&lt;-</span> <span class="fu">quickCluster</span><span class="op">(</span><span class="va">example_sce</span>,  assay.type <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># factor calculation</span></span>
<span><span class="va">size_factor</span> <span class="op">&lt;-</span> <span class="fu">calculateSumFactors</span><span class="op">(</span><span class="va">example_sce</span>, cluster <span class="op">=</span> <span class="va">cell_pooling</span>, min.mean <span class="op">=</span> <span class="fl">0.1</span>, assay.typ <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># normalize</span></span>
<span><span class="va">example_sce</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">example_sce</span>, size.factors <span class="op">=</span> <span class="va">size_factor</span> <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-4-choose-analysis-granularity">Step 4: Choose analysis granularity<a class="anchor" aria-label="anchor" href="#step-4-choose-analysis-granularity"></a>
</h2>
<p>The <em>seismic</em> framework requires a column of cell metadata to
specify the analysis granularity. In our analysis we care about cell
types and tissue-specific effects, so we combine them together.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_sce</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">$</span><span class="va">free_annotation</span><span class="op">)</span>, </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">$</span><span class="va">tissue</span>,<span class="st">"."</span>,<span class="va">example_sce</span><span class="op">$</span><span class="va">free_annotation</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">$</span><span class="va">tissue</span>,<span class="st">"."</span>,<span class="va">example_sce</span><span class="op">$</span><span class="va">cell_ontology_class</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-5-run-the-seismic-analysis">Step 5: Run the <em>seismic</em> analysis<a class="anchor" aria-label="anchor" href="#step-5-run-the-seismic-analysis"></a>
</h2>
<p>After going through the data processing steps we can then proceed
with the standard <em>seismic</em> analysis pipeline.</p>
<p>As a proof of concept we use the processed gene-level MAGMA z-scores
for type 2 diabetes included in the <em>seismic</em> package.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ylaboratory/seismic" class="external-link">seismicGWAS</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate specificity score</span></span>
<span><span class="va">sscore</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_specificity.html">calc_specificity</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">example_sce</span>, ct_label_col <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># map genes from mouse to human </span></span>
<span><span class="va">sscore_hsa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/translate_gene_ids.html">translate_gene_ids</a></span><span class="op">(</span><span class="va">sscore</span>, from <span class="op">=</span> <span class="st">"mmu_symbol"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate associations</span></span>
<span><span class="va">ct_association</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_ct_trait_associations.html">get_ct_trait_associations</a></span><span class="op">(</span>sscore <span class="op">=</span> <span class="va">sscore_hsa</span>, magma <span class="op">=</span> <span class="va">t2d_magma</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ct_association</span><span class="op">)</span></span></code></pre></div>
<pre class="console"><code><span><span class="co">##                                                            cell_type</span></span>
<span><span class="co">##                                                               &lt;char&gt;</span></span>
<span><span class="co">## 1:                                                Pancreas.beta cell</span></span>
<span><span class="co">## 2:                                        Pancreas.pancreatic A cell</span></span>
<span><span class="co">## 3:                                        Pancreas.pancreatic D cell</span></span>
<span><span class="co">## 4:            Large_Intestine.Lgr5- amplifying undifferentiated cell</span></span>
<span><span class="co">## 5:                            Large_Intestine.Goblet cell (Proximal)</span></span>
<span><span class="co">## 6: Brain_Non-Myeloid.excitatory neurons and some neuronal stem cells</span></span>
<span><span class="co">##          pvalue         FDR</span></span>
<span><span class="co">##           &lt;num&gt;       &lt;num&gt;</span></span>
<span><span class="co">## 1: 6.330263e-06 0.000481100</span></span>
<span><span class="co">## 2: 6.691778e-05 0.002542876</span></span>
<span><span class="co">## 3: 4.749954e-04 0.012033218</span></span>
<span><span class="co">## 4: 2.103253e-02 0.306608542</span></span>
<span><span class="co">## 5: 2.292316e-02 0.306608542</span></span>
<span><span class="co">## 6: 2.420594e-02 0.306608542</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qiliang Lai, Ruth Dannenfelser, Vicky Yao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
