<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A simple introduction to seismicGWAS • seismicGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="A simple introduction to seismicGWAS">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">seismicGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/seismicGWAS.html">Introduction</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-data-processing" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Data processing</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-data-processing">
<li><a class="dropdown-item" href="../articles/GWAS_processing.html">GWAS preprocessing</a></li>
    <li><a class="dropdown-item" href="../articles/scRNA-seq_processing.html">scRNA-seq preprocessing</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ylaboratory/seismic/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>A simple introduction to seismicGWAS</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ylaboratory/seismic/blob/master/vignettes/seismicGWAS.Rmd" class="external-link"><code>vignettes/seismicGWAS.Rmd</code></a></small>
      <div class="d-none name"><code>seismicGWAS.Rmd</code></div>
    </div>

    
    
<p>This vignette introduces the <code>seismicGWAS</code> package, an
implementation of the <em>seismic</em> framework for linking cell types
with traits using single-cell RNA-seq data and MAGMA summarized GWAS
data. We organize this vignette in two sections: Section 1 contains a
walk through using the sample data included with the
<code>seismicGWAS</code> package, and Section 2 walks through data
starting from a SingleCellExperiment object. In both cases, <a href="https://cncr.nl/research/magma/" class="external-link">MAGMA</a> is needed to produce
gene level GWAS summaries.</p>
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<p>First, install and load the seismic package. If you have not yet
installed <code>seismicGWAS</code>, use
<code>devtools::install(path_to_seismic_folder)</code> or
<code>devtools::install_github("ylaboratory/seismicGWAS")</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ylaboratory/seismic" class="external-link">seismicGWAS</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="section-1-running-seismic-on-the-included-sample-data">Section 1: running <em>seismic</em> on the included sample data<a class="anchor" aria-label="anchor" href="#section-1-running-seismic-on-the-included-sample-data"></a>
</h2>
<p><em>seismicGWAS</em> includes a preloaded small single cell RNAseq
dataset in the form of a SingleCellExperiment object
<code>tmfacs_sce_small</code></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmfacs_sce_small</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 23341 4000 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): logcounts</span></span>
<span><span class="co">#&gt; rownames(23341): 0610005C13Rik 0610007C21Rik ... l7Rn6</span></span>
<span><span class="co">#&gt;   zsGreen-transgene</span></span>
<span><span class="co">#&gt; rowData names(1): symbol</span></span>
<span><span class="co">#&gt; colnames(4000): M4.MAA000844.3_10_M.1.1 A18.MAA000388.3_11_M.1.1 ...</span></span>
<span><span class="co">#&gt;   M5.MAA000910.3_10_M.1.1 L2.MAA000563.3_10_M.1.1</span></span>
<span><span class="co">#&gt; colData names(17): nReads orig.ident ... sizeFactor cluster_name</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: RNA</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
<p>and two sample MAGMA files, one for Rheumatoid Arthritis (<a href="https://doi.org/10.1038/nature12873" class="external-link">Okada et al. 2013</a>)</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ra_magma</span><span class="op">)</span></span>
<span><span class="co">#&gt;     GENE CHR  START   STOP NSNPS NPARAM     N     ZSTAT       P</span></span>
<span><span class="co">#&gt; 1 148398   1 824993 889961     1      1 58284 -0.025069 0.51000</span></span>
<span><span class="co">#&gt; 2 339451   1 860967 911099     1      1 58284 -0.025069 0.51000</span></span>
<span><span class="co">#&gt; 3  84069   1 866872 920488    12      2 58284 -0.671780 0.74914</span></span>
<span><span class="co">#&gt; 4  26155   1 869583 929679    36      4 58284 -1.717100 0.95702</span></span>
<span><span class="co">#&gt; 5  84808   1 900579 952473    80      6 58284 -1.124900 0.86968</span></span>
<span><span class="co">#&gt; 6   9636   1 913847 959920    93      7 58284 -0.991290 0.83923</span></span></code></pre></div>
<p>and one for Type 2 Diabetes (<a href="https://doi.org/10.1038/s41467-018-04951-w%5D" class="external-link">Xue et
al. 2018</a>)</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">t2d_magma</span><span class="op">)</span></span>
<span><span class="co">#&gt;     GENE CHR   START    STOP NSNPS NPARAM      N    ZSTAT        P</span></span>
<span><span class="co">#&gt; 1 254173   1 1074286 1143315    21      3 572082  1.75840 0.039337</span></span>
<span><span class="co">#&gt; 2   8784   1 1128888 1177163     6      1 572870  1.93210 0.026671</span></span>
<span><span class="co">#&gt; 3  55210   1 1412523 1480067     9      1 569925  0.24436 0.403470</span></span>
<span><span class="co">#&gt; 4 339453   1 1460158 1510740    37      2 571069  0.14879 0.440860</span></span>
<span><span class="co">#&gt; 5  29101   1 1467053 1545262    39      2 571016  0.15608 0.437990</span></span>
<span><span class="co">#&gt; 6   9906   1 1646277 1712438    16      2 576212 -0.48124 0.684830</span></span></code></pre></div>
<div class="section level3">
<h3 id="calculating-cell-type-specificity-scores">Calculating cell type specificity scores<a class="anchor" aria-label="anchor" href="#calculating-cell-type-specificity-scores"></a>
</h3>
<p>The first step of <em>seismic</em> is to calculate cell type
specificity scores which will be used for subsequent cell type-trait
association calculations and influential gene analyses. Here we provide
the sample SingleCellExperiment object and column in the data containing
cell type labels <code>cluster_name</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmfacs_sscore</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_specificity.html">calc_specificity</a></span><span class="op">(</span><span class="va">tmfacs_sce_small</span>, ct_label_col<span class="op">=</span><span class="st">'cluster_name'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tmfacs_sscore</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span> <span class="co"># show a truncated view of the matrix</span></span>
<span><span class="co">#&gt;               Bladder.Bladder mesenchymal cell</span></span>
<span><span class="co">#&gt; 0610005C13Rik                       0.00000000</span></span>
<span><span class="co">#&gt; 0610007C21Rik                       0.05171198</span></span>
<span><span class="co">#&gt; 0610007L01Rik                       0.05053099</span></span>
<span><span class="co">#&gt; 0610007N19Rik                       0.10618537</span></span>
<span><span class="co">#&gt; 0610007P08Rik                       0.04881961</span></span>
<span><span class="co">#&gt; 0610007P14Rik                       0.04976406</span></span>
<span><span class="co">#&gt;               Bladder.Luminal bladder epithelial cell</span></span>
<span><span class="co">#&gt; 0610005C13Rik                            0.000000e+00</span></span>
<span><span class="co">#&gt; 0610007C21Rik                            5.416472e-02</span></span>
<span><span class="co">#&gt; 0610007L01Rik                            5.907819e-02</span></span>
<span><span class="co">#&gt; 0610007N19Rik                            4.726793e-05</span></span>
<span><span class="co">#&gt; 0610007P08Rik                            6.890189e-02</span></span>
<span><span class="co">#&gt; 0610007P14Rik                            6.196749e-02</span></span>
<span><span class="co">#&gt;               Brain_Myeloid.microglial cell</span></span>
<span><span class="co">#&gt; 0610005C13Rik                  3.639054e-19</span></span>
<span><span class="co">#&gt; 0610007C21Rik                  3.768650e-02</span></span>
<span><span class="co">#&gt; 0610007L01Rik                  3.409995e-07</span></span>
<span><span class="co">#&gt; 0610007N19Rik                 4.422689e-185</span></span>
<span><span class="co">#&gt; 0610007P08Rik                  2.758088e-08</span></span>
<span><span class="co">#&gt; 0610007P14Rik                  2.225707e-17</span></span></code></pre></div>
<p>The original single cell experiment in the
<code>tmfacs_sce_small</code> was derived from Tabula Muris. To
associate with human GWAS traits we need to convert the gene
identifiers. For convenience <em>seismicGWAS</em> has a function for
translating between gene ids for a specificity score matrix. We change
here from mouse gene symbols to human entrez ids to match the MAGMA
data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmfacs_sscore_hsa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/translate_gene_ids.html">translate_gene_ids</a></span><span class="op">(</span><span class="va">tmfacs_sscore</span>, from<span class="op">=</span><span class="st">'mmu_symbol'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tmfacs_sscore_hsa</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span> <span class="co"># show a truncated view of the matrix</span></span>
<span><span class="co">#&gt; 6 x 3 Matrix of class "dgeMatrix"</span></span>
<span><span class="co">#&gt;        Bladder.Bladder mesenchymal cell Bladder.Luminal bladder epithelial cell</span></span>
<span><span class="co">#&gt; 81932                      2.542912e-04                              0.01671191</span></span>
<span><span class="co">#&gt; 81577                      5.003187e-04                              0.01487892</span></span>
<span><span class="co">#&gt; 8614                       9.234388e-02                              0.02736780</span></span>
<span><span class="co">#&gt; 256281                     3.213228e-07                              0.08150739</span></span>
<span><span class="co">#&gt; 221687                     0.000000e+00                              0.00000000</span></span>
<span><span class="co">#&gt; 346389                     0.000000e+00                              0.17272120</span></span>
<span><span class="co">#&gt;        Brain_Myeloid.microglial cell</span></span>
<span><span class="co">#&gt; 81932                   7.503793e-12</span></span>
<span><span class="co">#&gt; 81577                   3.645070e-10</span></span>
<span><span class="co">#&gt; 8614                    0.000000e+00</span></span>
<span><span class="co">#&gt; 256281                  7.970981e-07</span></span>
<span><span class="co">#&gt; 221687                  0.000000e+00</span></span>
<span><span class="co">#&gt; 346389                  0.000000e+00</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculating-cell-type-trait-associations">Calculating cell type trait associations<a class="anchor" aria-label="anchor" href="#calculating-cell-type-trait-associations"></a>
</h3>
<p>We can now get cell type-trait associations for the RA and Type 2
Diabetes data and plot the top 20 associations per trait. Default plots
show FDR corrected p-values plotted on a -log scale so that stronger
associations appear at the top of the boxplot.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate and plot the cell type-trait associations for RA</span></span>
<span><span class="va">ra</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_ct_trait_associations.html">get_ct_trait_associations</a></span><span class="op">(</span><span class="va">tmfacs_sscore_hsa</span>, <span class="va">ra_magma</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_top_associations.html">plot_top_associations</a></span><span class="op">(</span><span class="va">ra</span>, limit <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="seismicGWAS_files/figure-html/unnamed-chunk-8-1.png" width="384"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate and plot the cell type-trait associations for T2D</span></span>
<span><span class="va">t2d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_ct_trait_associations.html">get_ct_trait_associations</a></span><span class="op">(</span><span class="va">tmfacs_sscore_hsa</span>, <span class="va">t2d_magma</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_top_associations.html">plot_top_associations</a></span><span class="op">(</span><span class="va">t2d</span>, limit <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="seismicGWAS_files/figure-html/unnamed-chunk-9-1.png" width="384"></p>
</div>
<div class="section level3">
<h3 id="influential-gene-analysis">Influential gene analysis<a class="anchor" aria-label="anchor" href="#influential-gene-analysis"></a>
</h3>
<p>With <em>seismicGWAS</em> we can also look at the genes driving a
particular cell type-trait association, in a process we call influential
gene analysis. Here, we need to select a particular cell type (in our
example we choose pancreas beta cell) and a trait (we use Type 2
Diabetes again). It is important to note that influential gene
calculations are only meaningful for cell type-trait associations that
are significant. So we have chosen a cell type and trait that were found
to be significant in the cell type-trait associations above.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate influential genes</span></span>
<span><span class="va">ct</span> <span class="op">&lt;-</span> <span class="st">"Pancreas.beta cell"</span></span>
<span><span class="va">t2d_beta_inf_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_inf_genes.html">find_inf_genes</a></span><span class="op">(</span><span class="va">ct</span>, <span class="va">tmfacs_sscore_hsa</span>, <span class="va">t2d_magma</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">t2d_beta_inf_genes</span><span class="op">)</span></span>
<span><span class="co">#&gt;      gene specificity  zstat   dfbetas is_influential</span></span>
<span><span class="co">#&gt;    &lt;char&gt;       &lt;num&gt;  &lt;num&gt;     &lt;num&gt;         &lt;lgcl&gt;</span></span>
<span><span class="co">#&gt; 1:   3767   0.4410499 6.1866 0.1941221           TRUE</span></span>
<span><span class="co">#&gt; 2: 169026   0.4932408 5.5197 0.1886886           TRUE</span></span>
<span><span class="co">#&gt; 3:   2645   0.4184336 5.2990 0.1513507           TRUE</span></span>
<span><span class="co">#&gt; 4:   3651   0.8982126 3.2276 0.1386102           TRUE</span></span>
<span><span class="co">#&gt; 5:   6833   0.4991599 4.1322 0.1293991           TRUE</span></span>
<span><span class="co">#&gt; 6:   7466   0.2439394 7.5000 0.1256267           TRUE</span></span></code></pre></div>
<p>Note: Only genes with positive contributions to the cell type-trait
association are marked influential by default. While negative
contributions may be informative, they are not as directly interpretable
and are driven by genes from other cell types or are somewhat cell type
specific but not associated with the given trait.</p>
<p>Next, to aid in result interpretation we convert the Entrez gene id
to human-readable gene symbols using mouse genome wide annotation
database (org.Hs.edg.db) and AnnotationDbi.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># convert ids to symbols using org.mm.eg.db and AnnotationDbi</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span> <span class="co">#make sure it has been installed</span></span>
<span><span class="va">hsa.map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>,  keys <span class="op">=</span> <span class="va">t2d_beta_inf_genes</span><span class="op">$</span><span class="va">gene</span>, keytype <span class="op">=</span> <span class="st">"ENTREZID"</span>,</span>
<span>                                 column <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hsa.map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/stack.html" class="external-link">stack</a></span><span class="op">(</span><span class="va">hsa.map</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">hsa.map</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"symbol"</span>, <span class="st">"entrezid"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">t2d_beta_inf_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">t2d_beta_inf_genes</span>, <span class="va">hsa.map</span>, </span>
<span>                            by.x<span class="op">=</span><span class="st">'gene'</span>, by.y<span class="op">=</span><span class="st">'entrezid'</span><span class="op">)</span></span></code></pre></div>
<p>We can also use the gene symbols to label the influential points in
the influential gene plot.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot influential genes</span></span>
<span><span class="fu"><a href="../reference/plot_inf_genes.html">plot_inf_genes</a></span><span class="op">(</span><span class="va">t2d_beta_inf_genes</span>, gene_col <span class="op">=</span> <span class="st">'symbol'</span>, num_labels <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="seismicGWAS_files/figure-html/unnamed-chunk-12-1.png" width="384"></p>
</div>
</div>
<div class="section level2">
<h2 id="section-2-running-seismicgwas-on-tabula-muris-senis">Section 2: running <em>seismicGWAS</em> on Tabula Muris Senis<a class="anchor" aria-label="anchor" href="#section-2-running-seismicgwas-on-tabula-muris-senis"></a>
</h2>
<p>Generally, we assume that users will be able to construct and
appropriately normalize their single cell data before using
<code>seismicGWAS</code>. For these tasks we recommend using the
<code>scran</code> package.</p>
<p>In this section we illustrate how to run the <em>seismicGWAS</em>
package on a real world single cell dataset from Tabula Muris. For
convenience we use the <a href="https://github.com/hemberg-lab/scfind" class="external-link"><code>scFind</code></a>
package to grab tissue level SingleCellExperiment objects.</p>
<div class="section level3">
<h3 id="grabbing-tabula-muris-pancreas-data-with-scfind">Grabbing Tabula Muris pancreas data with scFind<a class="anchor" aria-label="anchor" href="#grabbing-tabula-muris-pancreas-data-with-scfind"></a>
</h3>
<p>For simplicity, we grab one tissue type’s worth of data from the
<code>scfind</code> package.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># if scFind is not installed can install with devtools</span></span>
<span><span class="co"># devtools::install_github("hemberg-lab/scfind")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hemberg-lab/scfind" class="external-link">scfind</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="va">sce_panc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="va">tmfacs</span><span class="op">[</span><span class="st">"Pancreas"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Explore the SCE object to find the relevant cell type column. If one
did not exist, or one did not exist with our desired specificity, we
could create it by clustering, combining, or using other metadata and
adding to the SCE object before continuing.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce_panc</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 1961 rows and 7 columns</span></span>
<span><span class="co">#&gt;                             mouse     well             cell_type1   tissue</span></span>
<span><span class="co">#&gt;                          &lt;factor&gt; &lt;factor&gt;               &lt;factor&gt; &lt;factor&gt;</span></span>
<span><span class="co">#&gt; B21.MAA000574.3_8_M.1.1     3_8_M      B21              NA        Pancreas</span></span>
<span><span class="co">#&gt; D12.MAA000574.3_8_M.1.1     3_8_M      D12              NA        Pancreas</span></span>
<span><span class="co">#&gt; B22.MAA000574.3_8_M.1.1     3_8_M      B22              NA        Pancreas</span></span>
<span><span class="co">#&gt; D13.MAA000574.3_8_M.1.1     3_8_M      D13              leukocyte Pancreas</span></span>
<span><span class="co">#&gt; C1.MAA000574.3_8_M.1.1      3_8_M      C1               NA        Pancreas</span></span>
<span><span class="co">#&gt; ...                           ...      ...                    ...      ...</span></span>
<span><span class="co">#&gt; I9.MAA001868.3_38_F.1.1    3_38_F      I9  pancreatic acinar cell Pancreas</span></span>
<span><span class="co">#&gt; K10.MAA001868.3_38_F.1.1   3_38_F      K10 NA                     Pancreas</span></span>
<span><span class="co">#&gt; M19.MAA001868.3_38_F.1.1   3_38_F      M19 pancreatic acinar cell Pancreas</span></span>
<span><span class="co">#&gt; O9.MAA001868.3_38_F.1.1    3_38_F      O9  NA                     Pancreas</span></span>
<span><span class="co">#&gt; P16.MAA001868.3_38_F.1.1   3_38_F      P16 NA                     Pancreas</span></span>
<span><span class="co">#&gt;                          subtissue     FACS      sex</span></span>
<span><span class="co">#&gt;                           &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;</span></span>
<span><span class="co">#&gt; B21.MAA000574.3_8_M.1.1   Exocrine   Viable        M</span></span>
<span><span class="co">#&gt; D12.MAA000574.3_8_M.1.1   Exocrine   Viable        M</span></span>
<span><span class="co">#&gt; B22.MAA000574.3_8_M.1.1   Exocrine   Viable        M</span></span>
<span><span class="co">#&gt; D13.MAA000574.3_8_M.1.1   Exocrine   Viable        M</span></span>
<span><span class="co">#&gt; C1.MAA000574.3_8_M.1.1    Exocrine   Viable        M</span></span>
<span><span class="co">#&gt; ...                            ...      ...      ...</span></span>
<span><span class="co">#&gt; I9.MAA001868.3_38_F.1.1   Exocrine   Viable        F</span></span>
<span><span class="co">#&gt; K10.MAA001868.3_38_F.1.1  Exocrine   Viable        F</span></span>
<span><span class="co">#&gt; M19.MAA001868.3_38_F.1.1  Exocrine   Viable        F</span></span>
<span><span class="co">#&gt; O9.MAA001868.3_38_F.1.1   Exocrine   Viable        F</span></span>
<span><span class="co">#&gt; P16.MAA001868.3_38_F.1.1  Exocrine   Viable        F</span></span></code></pre></div>
<p>The <code>cell_type1</code> field contains the cell annotations that
we can use for calculating cell specificity scores. Right now there are
some <code>NA</code> values for field. Lets remove them before
continuing.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove cells with NA from the SCE object</span></span>
<span><span class="va">sce_panc_filt</span> <span class="op">&lt;-</span> <span class="va">sce_panc</span><span class="op">[</span>,<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce_panc</span><span class="op">)</span><span class="op">$</span><span class="va">cell_type1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce_panc_filt</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 1327 rows and 8 columns</span></span>
<span><span class="co">#&gt;                             mouse     well             cell_type1   tissue</span></span>
<span><span class="co">#&gt;                          &lt;factor&gt; &lt;factor&gt;               &lt;factor&gt; &lt;factor&gt;</span></span>
<span><span class="co">#&gt; D13.MAA000574.3_8_M.1.1     3_8_M      D13 leukocyte              Pancreas</span></span>
<span><span class="co">#&gt; D14.MAA000574.3_8_M.1.1     3_8_M      D14 pancreatic acinar cell Pancreas</span></span>
<span><span class="co">#&gt; C5.MAA000574.3_8_M.1.1      3_8_M      C5  pancreatic acinar cell Pancreas</span></span>
<span><span class="co">#&gt; A3.MAA000574.3_8_M.1.1      3_8_M      A3  pancreatic PP cell     Pancreas</span></span>
<span><span class="co">#&gt; C11.MAA000574.3_8_M.1.1     3_8_M      C11 pancreatic acinar cell Pancreas</span></span>
<span><span class="co">#&gt; ...                           ...      ...                    ...      ...</span></span>
<span><span class="co">#&gt; M16.MAA001868.3_38_F.1.1   3_38_F      M16 pancreatic acinar cell Pancreas</span></span>
<span><span class="co">#&gt; M17.MAA001868.3_38_F.1.1   3_38_F      M17 pancreatic acinar cell Pancreas</span></span>
<span><span class="co">#&gt; P15.MAA001868.3_38_F.1.1   3_38_F      P15 pancreatic acinar cell Pancreas</span></span>
<span><span class="co">#&gt; I9.MAA001868.3_38_F.1.1    3_38_F      I9  pancreatic acinar cell Pancreas</span></span>
<span><span class="co">#&gt; M19.MAA001868.3_38_F.1.1   3_38_F      M19 pancreatic acinar cell Pancreas</span></span>
<span><span class="co">#&gt;                          subtissue     FACS      sex sizeFactor</span></span>
<span><span class="co">#&gt;                           &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; D13.MAA000574.3_8_M.1.1   Exocrine   Viable        M  1.9765270</span></span>
<span><span class="co">#&gt; D14.MAA000574.3_8_M.1.1   Exocrine   Viable        M  1.2732105</span></span>
<span><span class="co">#&gt; C5.MAA000574.3_8_M.1.1    Exocrine   Viable        M  0.7804386</span></span>
<span><span class="co">#&gt; A3.MAA000574.3_8_M.1.1    Exocrine   Viable        M  0.0969612</span></span>
<span><span class="co">#&gt; C11.MAA000574.3_8_M.1.1   Exocrine   Viable        M  1.4387834</span></span>
<span><span class="co">#&gt; ...                            ...      ...      ...        ...</span></span>
<span><span class="co">#&gt; M16.MAA001868.3_38_F.1.1  Exocrine   Viable        F  0.1675565</span></span>
<span><span class="co">#&gt; M17.MAA001868.3_38_F.1.1  Exocrine   Viable        F  0.0712616</span></span>
<span><span class="co">#&gt; P15.MAA001868.3_38_F.1.1  Exocrine   Viable        F  0.1585532</span></span>
<span><span class="co">#&gt; I9.MAA001868.3_38_F.1.1   Exocrine   Viable        F  0.2143557</span></span>
<span><span class="co">#&gt; M19.MAA001868.3_38_F.1.1  Exocrine   Viable        F  0.2392390</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculating-cell-type-trait-associations-1">Calculating cell type-trait associations<a class="anchor" aria-label="anchor" href="#calculating-cell-type-trait-associations-1"></a>
</h3>
<p>Now we can calculate the cell type specificity scores we will need
for other downstream functions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate cell specificity</span></span>
<span><span class="va">panc_sscore</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_specificity.html">calc_specificity</a></span><span class="op">(</span><span class="va">sce_panc_filt</span>, ct_label_col<span class="op">=</span><span class="st">'cell_type1'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">panc_sscore</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span> <span class="co"># show a truncated view of the matrix</span></span>
<span><span class="co">#&gt;               endothelial cell    leukocyte pancreatic A cell</span></span>
<span><span class="co">#&gt; 0610007C21Rik     1.925564e-02 8.198702e-08      1.996716e-01</span></span>
<span><span class="co">#&gt; 0610007L01Rik     2.031098e-03 8.280322e-06      1.576197e-01</span></span>
<span><span class="co">#&gt; 0610007N19Rik     4.378839e-13 0.000000e+00      5.664380e-19</span></span>
<span><span class="co">#&gt; 0610007P08Rik     3.511300e-13 2.826381e-05      1.804820e-01</span></span>
<span><span class="co">#&gt; 0610007P14Rik     1.026578e-03 6.615702e-06      1.918313e-01</span></span>
<span><span class="co">#&gt; 0610007P22Rik     1.052606e-04 8.502420e-03      1.250415e-01</span></span>
<span><span class="co">#&gt;               pancreatic acinar cell</span></span>
<span><span class="co">#&gt; 0610007C21Rik          8.149700e-216</span></span>
<span><span class="co">#&gt; 0610007L01Rik           3.567238e-43</span></span>
<span><span class="co">#&gt; 0610007N19Rik           6.355130e-02</span></span>
<span><span class="co">#&gt; 0610007P08Rik           3.894345e-12</span></span>
<span><span class="co">#&gt; 0610007P14Rik          5.294229e-107</span></span>
<span><span class="co">#&gt; 0610007P22Rik           7.515759e-21</span></span></code></pre></div>
<p>Since we are using a subset of Tabula Muris it is time again to
convert mouse gene symbols to human entrez ids before continuing so that
we can match the identifiers in the human Type 2 Diabetes GWAS data.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">panc_sscore_hsa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/translate_gene_ids.html">translate_gene_ids</a></span><span class="op">(</span><span class="va">panc_sscore</span>, from<span class="op">=</span><span class="st">'mmu_symbol'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">panc_sscore_hsa</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span> <span class="co"># show a truncated view of the matrix</span></span>
<span><span class="co">#&gt; 6 x 4 Matrix of class "dgeMatrix"</span></span>
<span><span class="co">#&gt;        endothelial cell    leukocyte pancreatic A cell pancreatic acinar cell</span></span>
<span><span class="co">#&gt; 81932       0.004922072 0.000000e+00      2.088431e-01           6.213640e-04</span></span>
<span><span class="co">#&gt; 81577       0.119625749 1.955628e-16      1.127160e-01           3.838398e-31</span></span>
<span><span class="co">#&gt; 8614        0.000000000 0.000000e+00      2.420464e-05           1.766824e-01</span></span>
<span><span class="co">#&gt; 256281      0.004280693 9.500268e-03      1.938377e-01           8.279519e-95</span></span>
<span><span class="co">#&gt; 221687      0.000000000 0.000000e+00      1.659393e-37           0.000000e+00</span></span>
<span><span class="co">#&gt; 346389      0.000000000 0.000000e+00      1.067150e-07           1.468089e-05</span></span></code></pre></div>
<p>Finally, we can calculate cell type-trait associations as before.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">panc_t2d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_ct_trait_associations.html">get_ct_trait_associations</a></span><span class="op">(</span><span class="va">panc_sscore_hsa</span>, <span class="va">t2d_magma</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_top_associations.html">plot_top_associations</a></span><span class="op">(</span><span class="va">panc_t2d</span>, limit <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="seismicGWAS_files/figure-html/unnamed-chunk-18-1.png" width="384"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qiliang Lai, Ruth Dannenfelser, Vicky Yao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
